Several little bug fixes

drzowie opened the pull request:
13-Sep-2012

Dima,

See my email directly to you.  Lots of work and improvements, including (on the latest push) mouse input.

---

Hey, Dima,

I've been desultorily improving my copy of PDL::Graphics::Gnuplot for a while.  Just realized we've probably gotten out of sync.  The latest includes FITS file support and better documentation, includes some little fixes of edge cases in the parsing (e.g. 'label=>["My label", at=>[50,120]]' works; before, you would have had to say 'label=>["My label", at=>"50,120"]'.  

I also added some logic to allow plot options at the end -- so you can (e.g.) 

  $w=gpwin();
  $w->line(xvals(50),xvals(50)**3/2500,{title=>"my plot"});

That turns out to be the way many folks want to write stuff:  data first, options later.

dkogan commented:
> On Wed, 4 Apr 2012 15:45:06 -0700
> drzowie
> reply@reply.github.com
> wrote:
> 
> Hey, Dima,
> 
> I've been desultorily improving my copy of PDL::Graphics::Gnuplot for
> a while.  Just realized we've probably gotten out of sync.  The
> latest includes FITS file support and better documentation, includes
> some little fixes of edge cases in the parsing (e.g. 'label=>["My
> label", at=>[50,120]]' works; before, you would have had to say
> 'label=>["My label", at=>"50,120"]'.  
> 
> I also added some logic to allow plot options at the end -- so you
> can (e.g.) 
>   $w=gpwin();
>   $w->line(xvals(50),xvals(50)**3/2500,{title=>"my plot"});
> 
> That turns out to be the way many folks want to write stuff:  data
> first, options later.

Hi Craig.

I haven't touched the code of the module at all recently, so we haven't
gotten out of sync at all. :) I do use the module quite a bit, in the
original form I had it in. I don't have a lot of time to work on new
features, and it's already feature-complete for what I need, so maybe
it makes sense for you to take over as the official keeper of the main
tree. Does that sound reasonable?

I regularly run some code that uses this module, and it would be highly
desirable for it to work, even using your latest stuff. I just tried
it, and there are a number of things that don't work with the new stuff:
1. I use the object-oriented interface. It looks like you now require
   ->new(terminal, {termopts}, {plotopts}). Was it difficult to allow
   plain ->new(plotopts) as before? I can imagine getting naming conflicts
   and such. 2. 'square' plot option doesn't seem to exist anymore; at
   least I get errors when I try to use it
2. Past these two, I haven't yet figured out why it's failing. Some of
   my plots complain that a scalar is being used as an array-ref, and
   others complain that the numbers of legends and curves don't match. I
   haven't dug too deeply here yet.

Other various observations:
1. Plots aren't interactive anymore. Even when you run the most basic
   parabola plot from the top of the manpage, I can't interactively zoom.
2. The demo.pl test suite doesn't pass for me. Is this because it does
   some stuff you don't support anymore, or is this a not-yet-solved thing?
3. Are you allowing both plot(data, options) AND plot(options, data)?
   That sounds magical :). What do you do if you get plot(data, options,
   data) ?

I don't mean to disparage the work you're doing at all. There're lots of
options, and different people all want to interact with the tool in
different ways, so supporting everything and satisfying everybody is
impossible. I'm happy to change the way I'm calling the functions in my
code to work with the newer stuff, once I read about the new API. I'd
like ->new to be able to take straight plot options as before, using
the default terminal; this is a 'like', not a requirement for me,
though. Things like interactivity are a must.

If you have some grand, long-term, strongly incompatible vision, then
we can split my old stuff off into PDL::Graphics::Gnuplot::Simple, or
something like that. Then THAT module is pretty much stable, and you
can be completely unchained from the old way of doing things.

Thoughts?
dima

drzowie commented:
Heh.

I really, really have to focus on a test suite, sorry it has dragged on.  For this last push, I had apparently gotten my branches mixed up and pulled an ancient copy before making that last mod.  The current version on github (my  main branch) now has all the goodies back in it from the last few revs.

All the best,
Craig

On Apr 7, 2012, at 3:02 AM, Dima Kogan wrote:

> > On Wed, 4 Apr 2012 15:45:06 -0700
> > drzowie
> > reply@reply.github.com
> > wrote:
> > 
> > Hey, Dima,
> > 
> > I've been desultorily improving my copy of PDL::Graphics::Gnuplot for
> > a while.  Just realized we've probably gotten out of sync.  The
> > latest includes FITS file support and better documentation, includes
> > some little fixes of edge cases in the parsing (e.g. 'label=>["My
> > label", at=>[50,120]]' works; before, you would have had to say
> > 'label=>["My label", at=>"50,120"]'.  
> > 
> > I also added some logic to allow plot options at the end -- so you
> > can (e.g.) 
> >  $w=gpwin();
> >  $w->line(xvals(50),xvals(50)**3/2500,{title=>"my plot"});
> > 
> > That turns out to be the way many folks want to write stuff:  data
> > first, options later.
> 
> Hi Craig.
> 
> I haven't touched the code of the module at all recently, so we haven't
> gotten out of sync at all. :) I do use the module quite a bit, in the
> original form I had it in. I don't have a lot of time to work on new
> features, and it's already feature-complete for what I need, so maybe
> it makes sense for you to take over as the official keeper of the main
> tree. Does that sound reasonable?
> 
> I regularly run some code that uses this module, and it would be highly
> desirable for it to work, even using your latest stuff. I just tried
> it, and there are a number of things that don't work with the new stuff:
> 1. I use the object-oriented interface. It looks like you now require
>    ->new(terminal, {termopts}, {plotopts}). Was it difficult to allow
>    plain ->new(plotopts) as before? I can imagine getting naming conflicts
>    and such. 2. 'square' plot option doesn't seem to exist anymore; at
>    least I get errors when I try to use it
> 2. Past these two, I haven't yet figured out why it's failing. Some of
>    my plots complain that a scalar is being used as an array-ref, and
>    others complain that the numbers of legends and curves don't match. I
>    haven't dug too deeply here yet.
> 
> Other various observations:
> 1. Plots aren't interactive anymore. Even when you run the most basic
>    parabola plot from the top of the manpage, I can't interactively zoom.
> 2. The demo.pl test suite doesn't pass for me. Is this because it does
>    some stuff you don't support anymore, or is this a not-yet-solved thing?
> 3. Are you allowing both plot(data, options) AND plot(options, data)?
>    That sounds magical :). What do you do if you get plot(data, options,
>    data) ?
> 
> I don't mean to disparage the work you're doing at all. There're lots of
> options, and different people all want to interact with the tool in
> different ways, so supporting everything and satisfying everybody is
> impossible. I'm happy to change the way I'm calling the functions in my
> code to work with the newer stuff, once I read about the new API. I'd
> like ->new to be able to take straight plot options as before, using
> the default terminal; this is a 'like', not a requirement for me,
> though. Things like interactivity are a must.
> 
> If you have some grand, long-term, strongly incompatible vision, then
> we can split my old stuff off into PDL::Graphics::Gnuplot::Simple, or
> something like that. Then THAT module is pretty much stable, and you
> can be completely unchained from the old way of doing things.
> 
> Thoughts?
> dima
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/dkogan/PDL-Graphics-Gnuplot/pull/2#issuecomment-5006910

drzowie commented:
Hi, Dima,

I just realized I didn't reply to your earlier e-mail, just the "various observations".  Here are some thoughts.

On Apr 7, 2012, at 11:02 AM, Dima Kogan wrote:

> Hi Craig.
> 
> I haven't touched the code of the module at all recently, so we haven't
> gotten out of sync at all. :) I do use the module quite a bit, in the
> original form I had it in. I don't have a lot of time to work on new
> features, and it's already feature-complete for what I need, so maybe
> it makes sense for you to take over as the official keeper of the main
> tree. Does that sound reasonable?

That sounds reasonable to me, if you'd like to hand over the CPAN module or something - but I don't want to interfere with the fine work you've done, and also (like you) I tend to do work in occasional bursts, then ignore things for a long time.

> I regularly run some code that uses this module, and it would be highly
> desirable for it to work, even using your latest stuff. I just tried
> it, and there are a number of things that don't work with the new stuff:

I think most of the inconsistencies are now fixed.  You may still see a few minor issues, which I'll elaborate below.

> 1. I use the object-oriented interface. It looks like you now require
>    ->new(terminal, {termopts}, {plotopts}). Was it difficult to allow
>    plain ->new(plotopts) as before? I can imagine getting naming conflicts
>    and such. 

Yes, that is exactly the issue.  There are naming conflicts between some of the plot options and terminal options.  That is exacerbated by the simplistic abbrev scheme I adopted, which makes it awkward to disambiguate abbbrevs between two sets of options.  

> 1. 'square' plot option doesn't seem to exist anymore; at
>    least I get errors when I try to use it

That is fixed now.  I implemented "justify" by analogy to the PGPLOT option of the same name, but forgot to implement your "square".  

> 1. Past these two, I haven't yet figured out why it's failing. Some of
>    my plots complain that a scalar is being used as an array-ref, and
>    others complain that the numbers of legends and curves don't match. I
>    haven't dug too deeply here yet.

I'm slightly less forgiving in how I parse arguments, in part because I wanted (the user) to be able to pass arguments around in hash refs as well as in lists.  That means things like your frequent construct "legend=>'foo',legend=>'bar'" don't work, because there's a duplicated key.  I patched that up by allowing the construct "legend=>['foo','bar']" to pass in multiple legends.  I hope you won't see that as too evil.

There were several bugs in which a parser generated a scalar and an emitter expected an array ref; I think those are gone.

> Other various observations:
> 1. Plots aren't interactive anymore. Even when you run the most basic
>    parabola plot from the top of the manpage, I can't interactively zoom.

Ah.  Yes, this came about because I stupidly put a "reset" at the bottom of every plot command.  That is fixed.

> 1. The demo.pl test suite doesn't pass for me. Is this because it does
>    some stuff you don't support anymore, or is this a not-yet-solved thing?

Most of that was because of the change in multilegend handling.  I've tweaked the demo.pl to match the newer calling form.

> 1. Are you allowing both plot(data, options) AND plot(options, data)?
>    That sounds magical :). What do you do if you get plot(data, options,
>    data) ?

Curve options for a given curve have to come before the data for that curve, and if there are trailing plot options they have to be in a hash ref.  Otherwise everything "just works".  So you can pass in (plot options as list or hash ref), followed by zero or more of (curve options as list or hash ref, followed by data), followed by (optional plot options as hash ref).  Data,options,data would plot the first data with current object plot options and either globalwith or line (depending on whether globalwith was set).

There are some more nice things: replot() works now.  

> I don't mean to disparage the work you're doing at all. There're lots of
> options, and different people all want to interact with the tool in
> different ways, so supporting everything and satisfying everybody is
> impossible. I'm happy to change the way I'm calling the functions in my
> code to work with the newer stuff, once I read about the new API. I'd
> like ->new to be able to take straight plot options as before, using
> the default terminal; this is a 'like', not a requirement for me,
> though. Things like interactivity are a must.

Thanks for the kind words.  

Requiring a terminal type went hand-in-hand with allowing terminal options in the constructor - you can't parse terminal options until you know what terminal you've got.  But setting only plot options should be a valid option.  That is not in the tree just now, but if you felt like hacking it in I'd be happy :-)

> If you have some grand, long-term, strongly incompatible vision, then
> we can split my old stuff off into PDL::Graphics::Gnuplot::Simple, or
> something like that. Then THAT module is pretty much stable, and you
> can be completely unchained from the old way of doing things.

I don't think my vision is strongly incompatible with yours.  The current tree (with testing and bug fixes) is broadly compatible with your original thin-layer interface.  My intent over the next couple of days is to hack in mouse input (for a project I need to do next week).  With your approval, I want to issue a CPAN release and announce it to the PDL list.   Beyond that, I anticipate only periodic itch-scratching.

All the best,
Craig
